source/mesh_refinement/isosurfaces.cc

```diff
@@ -0,0 +1,391 @@
+/*
+  Copyright (C) 2011 - 2019 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+
+#include <aspect/mesh_refinement/isosurfaces.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/fe/fe_values.h>
+#include <math.h>
+#include <algorithm>
+
+namespace aspect
+{
+  namespace MeshRefinement
+  {
+    namespace Internal
+    {
+      bool
+      Isosurface::values_are_in_range(const std::vector<double> values) const
+      {
+        // This assumes that all the vectors in isosurfaces already have the
+        // same length.
+        Assert(values.size() == min_values.size(),
+               ExcMessage("Internal error: Vector of values passed to the isosurface class function values_are_in_range, "
+                          "does not have the the correct size."));
+        for (unsigned int index = 0; index < values.size(); ++index)
+          {
+            if (values[index] < min_values[index] || values[index] > max_values[index])
+              {
+                // outside this isosurface, no need to search further
+                return false;
+              }
+          }
+        // If we made it this far, then we are inside the conditions, so return true.
+        return true;
+      }
+
+      Property::Property(const std::string &property_name,const std::vector<std::string> &available_compositions)
+      {
+        bool found = false;
+        if (property_name == "Temperature")
+          {
+            name = PropertyName::Temperature;
+            index = 0;
+            found = true;
+          }
+        else if (property_name == "background")
+          {
+            name = PropertyName::Background;
+            index = 0;
+            found = true;
+          }
+        else
+          {
+            auto p = std::find(available_compositions.begin(), available_compositions.end(), property_name);
```

Comment:
Hm, what if I call one of my compositional fields "background"?

Reply:
I wasn't really aware of it, but @lhy11009 pointed out that this name seems to be already kind of reserved by the `parse_map_to_double_array` function. I have added an assert  in the parse parameters. I am open to other solutions. @lhy11009, what do you think?

Reply:
Yes, we have already reserved that for phase changes. I think it would be better to make it so here.